Win32-network/src/System/Win32/Async/IOManager.hsc

```diff
@@ -0,0 +1,274 @@
+                         -- ^ handle of a completion port
+                         -> IO ()
+dequeueCompletionPackets iocp@(IOCompletionPort port) =
+    alloca $ \numBytesPtr ->
+      alloca $ \lpCompletionKey ->
+        alloca $ \lpOverlappedPtr -> do
+      -- make 'GetQueuedCompletionStatus' system call which dequeues
+      -- a packet from io completion packet.  We use 'maxBound' as the timeouts
+      -- for dequeueing completion packets from IOCP.  The thread that runs
+      -- 'dequeueCompletionPackets' is ment to run for the entire execution time
+      -- of an appliction.
+      gqcsResult <- c_GetQueuedCompletionStatus port numBytesPtr lpCompletionKey lpOverlappedPtr maxBound
+      errorCode <- Win32.getLastError
+      lpOverlapped <- peek lpOverlappedPtr
+      completionKey <- peek lpCompletionKey
+      let gqcsOverlappedIsNull = lpOverlapped == nullPtr
+
+      -- gqcsIODataPtr was allocated by 'withIODataPtr' or
+      -- 'wsaWaitForCompletion', and we are responsible to deallocate it but
+      -- only if 'gqcsCompletionKey' is 'True'.
+
+      if | gqcsOverlappedIsNull && errorCode == eRROR_ABANDONED_WAIT_0 ->
+           -- this is triggered by closing IO completion 'HANDLE' while there
+           -- are outstanding requests (this is documented
+           -- <https://docs.microsoft.com/en-us/windows/win32/api/ioapiset/nf-ioapiset-getqueuedcompletionstatus#remarks>)
+              return ()
+         | gqcsOverlappedIsNull && errorCode == eRROR_INVALID_HANDLE ->
+           -- If we don't terminate the dequeueing thread, some of the tests
+           -- will take 10x-100x more time to complete:
+           --
+           -- - 'Async.Handle.async reads and writes'
+           -- - 'Async.Handle.PingPongPipelined test'
+           -- - 'Async.Socket.PingPong test'
+           -- - 'Async.Socket.PingPongPipelined test'
+           --
+           -- or not terminate at all within reasonable timeout:
+           --
+           -- - 'Async.Handle.PingPong test'
+           --
+           -- Note that in the test setup, we start io manager thread for each
+           -- test case.  If we don't terminate it when the iocp port is closed
+           -- we might end up with a busy loop which runs expensive system
+           -- calls that fail.  This could explain the slowdown.
+           --
+           -- In every observed case both values:
+           --
+           -- - 'gqcsResult'
+           -- - 'gqcsCompletionKey'
+           --
+           -- where false.
+           --
+               return ()
+         | gqcsOverlappedIsNull ->
+           -- 'GetLastCompletionStatus' has not dequeued any completion packet
+           -- from the completion port.  Must be the first clause, since if
+           -- this is not true we cannot trust other arguments.
+              throwIO NullOverlappedPointer
+         | completionKey /= magicCompletionKey ->
+            dequeueCompletionPackets iocp
```

Comment:
Meaning we ignore it and carry on. Lets say so.

Note also the comment about recursive calls below applies here too.